kube-client/src/client/client_ext.rs

```diff
@@ -42,9 +45,90 @@ pub struct Cluster;
 /// You can create this directly, or convert `From` a `String` / `&str`, or `TryFrom` an `k8s_openapi::api::core::v1::Namespace`
 pub struct Namespace(String);
 
+/// Referenced object name resolution
+pub trait ObjectRef<K>: ObjectUrl<K> {
+    fn name(&self) -> Option<&str>;
+}
+
+/// Reference resolver for a specified namespace
+pub trait NamespacedRef<K>: ObjectUrl<K> {
+    /// Resolve reference in the provided namespace
+    fn within(&self, namespace: String) -> impl ObjectRef<K>;
+}
+
+impl<K> ObjectUrl<K> for ObjectReference
+where
+    K: Resource,
+{
+    fn url_path(&self) -> String {
+        url_path(
+            &ApiResource::from_gvk(&self.clone().into()),
+            self.namespace.clone(),
+        )
+    }
+}
+
+impl<K> ObjectRef<K> for ObjectReference
+where
+    K: Resource,
+{
+    fn name(&self) -> Option<&str> {
+        self.name.as_deref()
+    }
+}
+
+impl<K> NamespacedRef<K> for ObjectReference
+where
+    K: Resource,
+    K::Scope: NamespaceScope,
+{
+    fn within(&self, namespace: String) -> impl ObjectRef<K> {
+        Self {
+            namespace: namespace.into(),
+            ..self.clone()
+        }
+    }
+}
+
+impl<K> ObjectUrl<K> for OwnerReference
```

Comment:
I think there are 2 more types that effectively we could implement for free as they generally have enough information:
- [`TypedObjectReference`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.TypedObjectReference.html)
- [`CrossVersionObjectReference`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/autoscaling/v2/struct.CrossVersionObjectReference.html)

perhaps good to do one/two of those to show the system is extensible enough.